javascript/ql/src/semmle/javascript/frameworks/Templating.qll

```diff
@@ -48,4 +48,665 @@ module Templating {
+
+  private TemplateSyntax getOwnTemplateSyntaxInFolder(Folder f) {
+    exists(PackageDependencies deps |
+      deps.getADependency(result.getAPackageName(), _) and
+      f = deps.getFile().getParentContainer()
+    )
+  }
+
+  private TemplateSyntax getTemplateSyntaxInFolder(Folder f) {
+    result = getOwnTemplateSyntaxInFolder(f)
+    or
+    not exists(getOwnTemplateSyntaxInFolder(f)) and
+    result = getTemplateSyntaxInFolder(f.getParentContainer())
+  }
+
+  private TemplateSyntax getTemplateSyntaxFromInstantiation(TemplateFile file) {
+    result = any(TemplateInstantiaton inst | inst.getTemplateFile() = file).getTemplateSyntax()
+  }
+
+  /**
+   * Gets a template syntax likely to be used in the given file.
+   */
+  TemplateSyntax getLikelyTemplateSyntax(TemplateFile file) {
+    result = getTemplateSyntaxFromInstantiation(file)
+    or
+    not exists(getTemplateSyntaxFromInstantiation(file)) and
+    result.getAFileExtension() = file.getExtension()
+    or
+    not exists(getTemplateSyntaxFromInstantiation(file)) and
+    not file.getExtension() = any(TemplateSyntax s).getAFileExtension() and
+    result = getTemplateSyntaxInFolder(file.getParentContainer())
+  }
+
+  /** A step through the `safe` pipe, which bypasses HTML escaping. */
+  private class SafePipeStep extends TaintTracking::SharedTaintStep {
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      exists(DataFlow::CallNode call |
+        call = getAPipeCall("safe") and
+        pred = call.getArgument(0) and
+        succ = call
+      )
+    }
+  }
+
+  /**
+   * An EJS-style `include` call within a template tag, such as `<%- include(file, { params }) %>`.
+   */
+  private class EjsIncludeCallInTemplate extends TemplateInstantiaton::Range, DataFlow::CallNode {
+    EjsIncludeCallInTemplate() {
+      exists(TemplatePlaceholderTag tag |
+        tag.getRawText().regexpMatch("(?s)<%-.*") and
+        this = tag.getInnerTopLevel().getAVariableUse("include").getACall()
+      )
+    }
+
+    /** Gets a data flow node that refers to the instantiated template, if any. */
+    override DataFlow::SourceNode getOutput() { result = this }
+
+    /** Gets a data flow node that refers a template file to be instantiated, if any. */
+    override DataFlow::Node getTemplateFileNode() { result = getArgument(0) }
+
+    /** Gets a data flow node that refers to an object whose properties become variables in the template. */
+    override DataFlow::Node getTemplateParamsNode() { result = getArgument(1) }
+  }
+
+  /** The `include` function, seen as an API node, so we can treat it as a template instantiation. */
+  private class IncludeFunctionAsEntryPoint extends API::EntryPoint {
+    IncludeFunctionAsEntryPoint() { this = "IncludeFunctionAsEntryPoint" }
+
+    override DataFlow::SourceNode getAUse() {
+      result = any(TemplatePlaceholderTag tag).getInnerTopLevel().getAVariableUse("include")
+    }
+
+    override DataFlow::Node getARhs() { none() }
+  }
```

Comment:
I can see that this class is needed, because the tests fail if I remove it.   

But I'm not sure where it's needed.  
I can't get it to fit with the `API::Node` usage in `getTemplateInput`, but maybe that's me missing something. 
Could you elaborate in which class/predicate this `API::EntryPoint` is needed?

Reply:
It's needed for `EjsIncludeCallInTemplate`. By ensuring that the callee has an API node, the arguments to call have API nodes as well, which is needed in `getTemplateParamsNode`, which is needed in `getTemplateInput`.

Using `EjsIncludeCallInTemplate` directly in the definition of the entry point leads to negative recursion, unfortunately.

Reply:
Great, could you add a reference to `EjsIncludeCallInTemplate` in the qldoc?